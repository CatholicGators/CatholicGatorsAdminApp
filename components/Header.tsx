import React from 'react';

import { withStyles, createStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';
import AccountCircle from '@material-ui/icons/AccountCircle';
import MenuItem from '@material-ui/core/MenuItem';
import Menu from '@material-ui/core/Menu';
import Button from '@material-ui/core/Button';
import { User } from 'lib/withAuth';

const styles = createStyles({
    root: {
        flexGrow: 1,
    },
    grow: {
        flexGrow: 1,
    },
    menuButton: {
        marginLeft: -12,
        marginRight: 20,
    },
});

interface HeaderProps {
    user: User;
    classes: any;
}

class Header extends React.Component {
    state = {
        anchorEl: null,
    };

    constructor(public props: HeaderProps) {
        super(props);
    }

    handleMenu(event) {
        this.setState({ anchorEl: event.currentTarget });
    };

    handleClose() {
        this.setState({ anchorEl: null });
    };

    render() {
        const { user, classes } = this.props;
        const { anchorEl } = this.state;

        return (
            <div className={classes.root}>
                <AppBar position="static">
                    <Toolbar>
                        <IconButton className={classes.menuButton} color="inherit" aria-label="Menu">
                            <MenuIcon />
                        </IconButton>
                        <Typography variant="h6" color="inherit" className={classes.grow}>
                            Catholic Gators Admin
                        </Typography>
                        {user ? (
                            <div>
                                <IconButton
                                    aria-owns={anchorEl ? 'menu-appbar' : undefined}
                                    aria-haspopup="true"
                                    onClick={this.handleMenu.bind(this)}
                                    color="inherit"
                                >
                                    <AccountCircle />
                                </IconButton>
                                <Menu
                                    id="menu-appbar"
                                    anchorEl={anchorEl}
                                    anchorOrigin={{
                                        vertical: 'top',
                                        horizontal: 'right',
                                    }}
                                    transformOrigin={{
                                        vertical: 'top',
                                        horizontal: 'right',
                                    }}
                                    open={anchorEl}
                                    onClose={this.handleClose.bind(this)}
                                >
                                    <MenuItem onClick={this.handleClose.bind(this)}>Profile</MenuItem>
                                    <MenuItem onClick={this.handleClose.bind(this)}>My account</MenuItem>
                                </Menu>
                            </div>
                        ) : (
                            <Button color="inherit">Login</Button>
                        )}
                    </Toolbar>
                </AppBar>
            </div>
        );
    }
}

export default withStyles(styles)(Header);
