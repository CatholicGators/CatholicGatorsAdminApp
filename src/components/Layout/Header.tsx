import React from 'react';

import { withStyles, createStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';
import Avatar from '@material-ui/core/Avatar';
import MenuItem from '@material-ui/core/MenuItem';
import Menu from '@material-ui/core/Menu';
import Button from '@material-ui/core/Button';

import withAuthentication from '../Auth/withAuthentication';

const styles = createStyles({
    root: {
        flexGrow: 1,
    },
    grow: {
        flexGrow: 1,
    },
    menuButton: {
        marginLeft: -12,
        marginRight: 20,
    },
});

class Header extends React.Component {
    state = {
        anchorEl: null,
    };

    constructor(public props: any) {
        super(props);
    }

    handleMenu(event) {
        this.setState({ anchorEl: event.currentTarget });
    }

    handleClose() {
        this.setState({ anchorEl: null });
    }

    handleLogin() {
        this.props.googleSignIn();
    }

    handleLogout() {
        this.props.signOut();
        this.handleClose();
    }

    render() {
        const { user, classes, children } = this.props;
        const { anchorEl } = this.state;

        return (
            <div className={classes.root}>
                <AppBar position="static">
                    <Toolbar>
                        <IconButton className={classes.menuButton} color="inherit" aria-label="Menu">
                            <MenuIcon />
                        </IconButton>
                        <Typography variant="h6" color="inherit" className={classes.grow}>
                            Catholic Gators Admin
                        </Typography>
                        {user ? (
                            <div>
                                <IconButton
                                    aria-owns={anchorEl ? 'menu-appbar' : undefined}
                                    aria-haspopup="true"
                                    onClick={this.handleMenu.bind(this)}
                                    color="inherit"
                                >
                                    <Avatar src={user.photoURL} />
                                </IconButton>
                                <Menu
                                    id="menu-appbar"
                                    anchorEl={anchorEl}
                                    anchorOrigin={{
                                        vertical: 'top',
                                        horizontal: 'right',
                                    }}
                                    transformOrigin={{
                                        vertical: 'top',
                                        horizontal: 'right',
                                    }}
                                    open={!!anchorEl}
                                    onClose={this.handleClose.bind(this)}
                                >
                                    <MenuItem onClick={this.handleLogout.bind(this)}>Logout</MenuItem>
                                </Menu>
                            </div>
                        ) : (
                            <Button color="inherit" onClick={this.handleLogin.bind(this)}>Login</Button>
                        )}
                    </Toolbar>
                </AppBar>
                {children}
            </div>
        );
    }
}

export default withAuthentication(withStyles(styles)(Header));