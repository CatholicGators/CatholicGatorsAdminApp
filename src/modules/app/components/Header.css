.root, .grow {
    flex-grow: 1;
}

.menuButton {
    margin-left: -12;
    margin-right: 20;
}