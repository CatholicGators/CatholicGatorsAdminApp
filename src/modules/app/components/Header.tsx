import React from 'react';
import { connect } from 'react-redux';
import { createStyles, withStyles } from '@material-ui/core';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';
import Avatar from '@material-ui/core/Avatar';
import MenuItem from '@material-ui/core/MenuItem';
import Menu from '@material-ui/core/Menu';
import Button from '@material-ui/core/Button';
// import CircularProgress from '@material-ui/core/CircularProgress';

import { googleSignIn, signOut } from '../../../redux/actions/auth/authActions';

const styles = createStyles({
    root: {
        flexGrow: 1
    },
    grow: {
        flexGrow: 1
    },
    menuButton: {
        marginLeft: -12,
        marginRight: 20
    }
});

type Props = {
    user: any,
    signOut: () => any,
    googleSignIn: () => any
}

export class Header extends React.Component<Props> {
    state = {
        anchorEl: null,
    };

    constructor(public props) {
        super(props);
    }

    handleMenu(event) {
        this.setState({ anchorEl: event.currentTarget });
    }

    handleClose() {
        this.setState({ anchorEl: null });
    }

    handleLogin() {
        this.props.googleSignIn();
    }

    handleLogout() {
        this.props.signOut();
        this.handleClose();
    }

    profile(user) {
        const { anchorEl } = this.state;

        switch(user) {
            case undefined:
                return <div>Loading...</div>; // TODO: Replace this with CircularProgress

            case null:
                return <Button id="login-btn" color="inherit" onClick={this.handleLogin.bind(this)}>Login</Button>

            default:
                return <div>
                    <IconButton
                        id="avatar-btn"
                        aria-owns={anchorEl ? 'menu' : undefined}
                        aria-haspopup="true"
                        onClick={this.handleMenu.bind(this)}
                        color="inherit"
                    >
                        <Avatar src={user.photoURL} />
                    </IconButton>
                    <Menu
                        id="menu"
                        anchorEl={anchorEl}
                        anchorOrigin={{
                            vertical: 'top',
                            horizontal: 'right',
                        }}
                        transformOrigin={{
                            vertical: 'top',
                            horizontal: 'right',
                        }}
                        open={!!anchorEl}
                        onClose={this.handleClose.bind(this)}
                    >
                        <MenuItem id="logout" onClick={this.handleLogout.bind(this)}>Logout</MenuItem>
                    </Menu>
                </div>
        }
    }

    render() {
        const { user, classes } = this.props;
        // const { anchorEl } = this.state;

        return (
            <div className={classes ? classes.root : null}>
                <AppBar position="static">
                    <Toolbar>
                        <IconButton className={classes ? classes.menuButton : null} color="inherit" aria-label="Menu">
                            <MenuIcon />
                        </IconButton>
                        <Typography variant="h6" color="inherit" className={classes ? classes.grow : null}>
                            Catholic Gators Admin
                        </Typography>
                        {this.profile(user)}
                    </Toolbar>
                </AppBar>
            </div>
        );
    }
}

const mapStateToProps = state => ({
    user: state.auth.user
});

const mapDispatchToProps = dispatch => ({
    googleSignIn: () => dispatch(googleSignIn()),
    signOut: () => dispatch(signOut())
});

export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(Header));
