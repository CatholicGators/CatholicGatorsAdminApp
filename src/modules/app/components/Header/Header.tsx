import React from 'react';
import { connect } from 'react-redux';
import { withRouter, NavLink } from "react-router-dom";

import { createStyles, withStyles, Theme } from '@material-ui/core';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';
import VpnKey from '@material-ui/icons/VpnKey';
import FormatAlignLeft from '@material-ui/icons/FormatAlignLeft';
import Inbox from '@material-ui/icons/Inbox';

import { googleSignIn, signOut } from '../../../../redux/actions/auth/authActions';
import MobileDrawer from '../MobileDrawer/MobileDrawer';
import ToolbarAvatar from '../ToolbarAvatar/ToolbarAvatar';

const styles = (theme: Theme) => createStyles({
    root: {
        flexGrow: 1
    },
    grow: {
        flexGrow: 1
    },
    menuButton: {
        marginLeft: -12,
        marginRight: 20,
        [theme.breakpoints.up('md')]: {
            display: 'none'
        }
    },
    toolbar: {
        minHeight: '64px'
    },
    desktopMenuItems: {
        display: 'flex',
        [theme.breakpoints.down('sm')]: {
            display: 'none'
        }
    },
    navLinkDesktop: {
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        paddingLeft: 20,
        paddingRight: 20,
        color: 'inherit',
        textDecoration: 'none',
        '&:visited': {
            color: 'inherit'
        },
        '&:hover': {
            color: theme.palette.secondary.main,
            transition: 'color',
            transitionDuration: '200ms',
        }
    },
    navLinkTextDesktop: {
        color: 'inherit'
    },
    desktopSelected: {
        color: theme.palette.secondary.main
    }
});

type Props = {
    user: any,
    signOut: () => any,
    googleSignIn: () => any
}

type State = {
    drawerOpen: boolean
}

export type MenuLink = {
    text: string,
    href: string,
    icon: any,
    needsAuthorization: boolean
}
export const menuLinks: Array<MenuLink> = [
    {
        text: 'Contact Form',
        href: '/',
        icon: FormatAlignLeft,
        needsAuthorization: false
    },
    {
        text: 'My Contacts',
        href: '/my-contacts',
        icon: Inbox,
        needsAuthorization: false
    },
    {
        text: 'Admin',
        href: '/admin',
        icon: VpnKey,
        needsAuthorization: true
    }
];

export class Header extends React.Component<Props, State> {
    state = {
        drawerOpen: false
    };

    constructor(public props) {
        super(props);
    }

    toggleDrawer(isOpen: boolean){
      this.setState({
        drawerOpen: isOpen,
      });
    };

    render() {
        const { classes, user } = this.props;

        return (
            <div className={classes ? classes.root : null}>
                <AppBar position="static">
                    <Toolbar className={classes ? classes.toolbar : null}>
                        <IconButton
                            id='menu-btn'
                            className={classes ? classes.menuButton : null}
                            color="inherit"
                            aria-label="Menu"
                            onClick={() => this.toggleDrawer(true)}
                        >
                            <MenuIcon/>
                        </IconButton>
                        <Typography variant="h6" color="inherit" className={classes ? classes.grow : null}>
                            Catholic Gators Admin
                        </Typography>
                        <div className={classes ? classes.desktopMenuItems : null}>
                            { user ?
                                menuLinks
                                    .filter(link => !link.needsAuthorization || (user.isAdmin && user.isApproved))
                                    .map(link => 
                                            <NavLink
                                                exact
                                                key={link.text}
                                                to={link.href}
                                                className={classes ? classes.navLinkDesktop : null}
                                                activeClassName={classes ? classes.desktopSelected : null}
                                            >
                                                <Typography
                                                    className={classes ? classes.navLinkTextDesktop : null}
                                                >
                                                    {link.text}
                                                </Typography>
                                            </NavLink>
                                    )
                                : null
                            }
                            <ToolbarAvatar
                                user={user}
                                login={() => this.props.googleSignIn()}
                                logout={() => this.props.signOut()}
                            />
                        </div>
                    </Toolbar>
                </AppBar>
                <MobileDrawer
                    isOpen={this.state.drawerOpen}
                    user={user}
                    closeDrawer={() => this.toggleDrawer(false)}
                    menuLinks={menuLinks}
                    logout={() => this.props.signOut() }
                    login={() => this.props.googleSignIn() }
                    selectedPath={this.props.location.pathname}
                />
            </div>
        );
    }
}

const mapStateToProps = state => ({
    user: state.auth.user
});

const mapDispatchToProps = dispatch => ({
    googleSignIn: () => dispatch(googleSignIn()),
    signOut: () => dispatch(signOut())
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(Header)));
