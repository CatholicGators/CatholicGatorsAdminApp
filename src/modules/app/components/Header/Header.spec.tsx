import React from 'react'
import { shallow, mount } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'

import { Header, menuLinks } from './Header'

describe('Header', () => {
    let wrapper, props

    beforeEach(() => {
        props = {
            user: {
                isAdmin: true
            },
            location: {
                pathname: '/'
            },
            signOut: jest.fn(),
            googleSignIn: jest.fn()
        }
        wrapper = shallow(<Header {...props} />)
    })

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('should have the drawer not open by default', () => {
        expect(wrapper.state('drawerOpen')).toBe(false)
    })

    it('renders all links when user is an admin', () => {
        const dom = mount(
            <MemoryRouter>
                <div>
                    <Header {...props} />
                </div>
            </MemoryRouter>
        )

        menuLinks.forEach(link => expect(dom.exists(`NavLink[to='${link.href}']`)).toBe(true))
    })

    it('renders only unAuthorized links when user is not an admin', () => {
        props = {
            ...props,
            user: {
                ...props.user,
                isAdmin: false
            }
        }
        const dom = mount(
            <MemoryRouter>
                <div>
                    <Header {...props} />
                </div>
            </MemoryRouter>
        )

        menuLinks.forEach(link => expect(dom.exists(`NavLink[to='${link.href}']`)).toBe(!link.needsAuthorization))
    })
})