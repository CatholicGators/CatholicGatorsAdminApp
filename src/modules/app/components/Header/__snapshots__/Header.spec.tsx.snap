// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match snapshot 1`] = `
<div
  className="test"
>
  <WithStyles(ForwardRef(AppBar))
    position="static"
  >
    <WithStyles(ForwardRef(Toolbar))
      className="test"
    >
      <WithStyles(ForwardRef(IconButton))
        aria-label="Menu"
        className="test"
        color="inherit"
        id="menu-btn"
        onClick={[Function]}
      >
        <MenuIcon />
      </WithStyles(ForwardRef(IconButton))>
      <WithStyles(ForwardRef(Typography))
        className="test"
        color="inherit"
        variant="h6"
      >
        Catholic Gators Admin
      </WithStyles(ForwardRef(Typography))>
      <div
        className="test"
      >
        <NavLink
          activeClassName="test"
          className="test"
          exact={true}
          key="Contact Form"
          to="/"
        >
          <WithStyles(ForwardRef(Typography))
            className="test"
          >
            Contact Form
          </WithStyles(ForwardRef(Typography))>
        </NavLink>
        <NavLink
          activeClassName="test"
          className="test"
          exact={false}
          key="My Contacts"
          to="/my-contacts"
        >
          <WithStyles(ForwardRef(Typography))
            className="test"
          >
            My Contacts
          </WithStyles(ForwardRef(Typography))>
        </NavLink>
        <NavLink
          activeClassName="test"
          className="test"
          exact={false}
          key="Admin"
          to="/admin"
        >
          <WithStyles(ForwardRef(Typography))
            className="test"
          >
            Admin
          </WithStyles(ForwardRef(Typography))>
        </NavLink>
        <WithStyles(ToolbarAvatar)
          login={[Function]}
          logout={[Function]}
          user={
            Object {
              "isAdmin": true,
              "isApproved": true,
            }
          }
        />
      </div>
    </WithStyles(ForwardRef(Toolbar))>
  </WithStyles(ForwardRef(AppBar))>
  <WithStyles(MobileDrawer)
    closeDrawer={[Function]}
    isOpen={false}
    login={[Function]}
    logout={[Function]}
    menuLinks={
      Array [
        Object {
          "href": "/",
          "icon": Object {
            "$$typeof": Symbol(react.memo),
            "compare": null,
            "displayName": "FormatAlignLeftIcon",
            "muiName": "SvgIcon",
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          },
          "needsAuthorization": false,
          "text": "Contact Form",
        },
        Object {
          "href": "/my-contacts",
          "icon": Object {
            "$$typeof": Symbol(react.memo),
            "compare": null,
            "displayName": "InboxIcon",
            "muiName": "SvgIcon",
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          },
          "needsAuthorization": false,
          "text": "My Contacts",
        },
        Object {
          "href": "/admin",
          "icon": Object {
            "$$typeof": Symbol(react.memo),
            "compare": null,
            "displayName": "VpnKeyIcon",
            "muiName": "SvgIcon",
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          },
          "needsAuthorization": true,
          "text": "Admin",
        },
      ]
    }
    selectedPath="/"
    user={
      Object {
        "isAdmin": true,
        "isApproved": true,
      }
    }
  />
</div>
`;
